### PR TITLE
Refactor TypeAndDataManager.h

### PR DESCRIPTION
**Pull request by mehagar**
From Resharper C++. Rules used:

local variable may be const
public base class access specifier is redundant
c style cast used instead of a c++ cast
Use auto
Inclusion of deprecated c header 'x'
member function may be const
macro argument should be enclosed in parentheses
redundant else keyword
use emplace_back instead of push_back
use =default to define a trivial constructor
avoid repeating the return type from the declaration, use a braced initializer instead

### CODE REVIEW

File: source/include/TypeAndDataManager.h [hunk 1]
```diff
@@ -1011,7 +1010,6 @@ class DefaultPointerType : public PointerType
 {
  private:
     DefaultPointerType(TypeManagerRef typeManager, TypeRef type, void* pointer, PointerTypeEnum pointerType);
-    DefaultPointerType();
```
Comment (sanmut) [inline source/include/TypeAndDataManager.h]: Why not "DefaultPointerType() = delete;"?

File: source/include/TypeAndDataManager.h [hunk 2]
```diff
@@ -1111,18 +1110,16 @@ class TypedArrayCore
...
-    AQBlock1* RawBegin()            { return _pRawBufferBegin; }
+    void* RawObj() { VIREO_ASSERT(Rank() == 0); return RawBegin(); }  // some extra asserts fo  ZDAs
+    AQBlock1* RawBegin() { return _pRawBufferBegin; }
     template<typename CT> CT BeginAtAQ(IntIndex index) { return reinterpret_cast<CT>(RawBegin() + index); }
     BlockItr RawItr()               { return BlockItr(RawBegin(), ElementType()->TopAQSize(), Length()); }
 
- public:
     //! Array's type.
-    TypeRef Type()                  { return _typeRef; }
+    TypeRef Type() { return _typeRef; }
```
Comment (sanmut) [inline source/include/TypeAndDataManager.h]: It's ok here and no need to address this.. In the future, ,it is a good idea to do format/style changes (such as whitespace, newline) separately in a PR.

File: source/include/TypeAndDataManager.h [hunk 3]
```diff
@@ -1245,7 +1242,7 @@ class String : public TypedArray1D< Utf8Char >
     void AppendEscapeEncoded(const Utf8Char* source, IntIndex len);
 
     void InsertSubString(IntIndex position, SubString* str) {
-        Insert(position, (IntIndex)str->Length(), (Utf8Char*)str->Begin());
+        Insert(position, static_cast<IntIndex>(str->Length()), const_cast<Utf8Char*>(str->Begin()));
```
Comment (sanmut) [inline source/include/TypeAndDataManager.h]: I think const_cast<Utf8Char*>is not required here. Just remove const_cast and even C style cast and try "str->Begin()"?

File: source/include/TypeAndDataManager.h [hunk 4]
```diff
@@ -285,18 +285,17 @@ class TypeManager
     Int32   AQAlignment(Int32 size);
     static Int32   AlignAQOffset(Int32 offset, Int32 size) {
         if (size != 0) {
-            Int32 remainder  = offset % size;
+            const Int32 remainder  = offset % size;
```
Comment (spathiwa) [inline source/include/TypeAndDataManager.h]: Adding class const modifiers to member functions is great, but I'm not sure what the point of adding const modifiers to scalar arguments or local variables which are not used in a way that is _conceptually_ constant such as this.  It would be hard to be consistent with this, and I think it adds no value.

File: source/include/TypeAndDataManager.h [hunk 5]
```diff
@@ -1326,16 +1325,20 @@ class StackVar
 
 struct StringRefCmp {
     bool operator()(const StringRef& a, const StringRef &b) const {
-        Int32 cmp = memcmp(a->Begin(), b->Begin(), Min(a->Length(), b->Length()));
+        const Int32 cmp = memcmp(a->Begin(), b->Begin(), Min(a->Length(), b->Length()));
```
Comment (spathiwa) [inline source/include/TypeAndDataManager.h]: Again, not sure why this is at all helpful.

File: source/include/TypeAndDataManager.h (hunk 4 above)
Comment (mehagar) [reply source/include/TypeAndDataManager.h]: I agree. I'll remove uses of const for local variables.

File: source/include/TypeAndDataManager.h (hunk 1 above)
Comment (mehagar) [reply source/include/TypeAndDataManager.h]: Because having a non default constructor declared already means that a default constructor won't be generated.

File: source/include/TypeAndDataManager.h [hunk 6]
```diff
@@ -1329,13 +1328,17 @@ struct StringRefCmp {
         Int32 cmp = memcmp(a->Begin(), b->Begin(), Min(a->Length(), b->Length()));
         if (cmp < 0) {
             return true;
-        } else if (cmp > 0) {
+        }
```
Comment (spathiwa) [inline source/include/TypeAndDataManager.h]: Please undo the changes to this function.  Leave the 'else' statements.
This is a good example of a situation where the code reads much better with them, regardless of whether they are 'redundant'.
